File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
    dummy := &ListNode{Next: head}
    fast := dummy
    slow := dummy

    for i:= 0; i < n; i++ {
        fast = fast.Next
    }

    for fast.Next != nil {
        slow = slow.Next
        fast = fast.Next
    }

    slow.Next = slow.Next.Next

    return dummy.Next
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int nodeLength = 1;
        ListNode current = head;
        while (current.next != null) {
            current = current.next;
            nodeLength++;
        }

        if (n >= nodeLength) {
            return head.next;
        }

        ListNode previousNode = head;
        current = previousNode.next;
        int idx = 0;

        while (idx < nodeLength - n - 1 && current != null) {
            idx++;
            previousNode = current;
            current = previousNode.next;
        }

        previousNode.next = current != null ? current.next : null;

        return head;
    }
}